x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java

```diff
@@ -84,27 +88,73 @@ public String toString() {
         return getClass().getSimpleName() + "{privileges=" + permissions + "}";
     }
 
+    public Set<String> getApplicationNames() {
+        return permissions.stream()
+            .map(e -> e.privilege.getApplication())
+            .collect(Collectors.toSet());
+    }
+
+    public Set<ApplicationPrivilege> getPrivileges(String application) {
+        return permissions.stream()
+            .filter(e -> application.equals(e.privilege.getApplication()))
+            .map(e -> e.privilege)
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns a set of resource patterns that are permitted for the provided privilege.
+     * The returned set may include patterns that overlap (e.g. "object/*" and "object/1") and may
+     * also include patterns that are defined again a more permissive privilege.
+     * e.g. If a permission grants
+     * <ul>
+     *     <li>"my-app", "read", [ "user/*" ]</li>
+     *     <li>"my-app", "all", [ "user/kimchy", "config/*" ]</li>
+     * </ul>
+     * Then <code>getResourcePatterns( myAppRead )</code> would return <code>"user/*", "user/kimchy", "config/*"</code>.
+     */
+    public Set<String> getResourcePatterns(ApplicationPrivilege privilege) {
+        return permissions.stream()
+            .filter(e -> e.matchesPrivilege(privilege))
+            .map(e -> e.resourceNames)
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
```

Comment:
It looks to me that you went to great lengths to display app privileges in nicely formatted JSON.
What do you think if you also do *resource-wildcard-compression* here? I mean, to go through
the flatted array and eliminate resources that are a subset of another, eg `user/kimchy` is a subset of `user/*` so do not print it.
Again, I would not suggest this if you hadn't already reached up to here! Kibana will be thrilled!

Reply:
This is trickier than it sounds.
The Automatons are (in some cases) already unioned, so calculating subsets is tricky.

In the example from this javadoc:
```
"read", [ "user/*" ]
"all", [ "user/kimchy", "config/*" ]
```
the two resources for `all` are already merged into a single automaton, so there's nothing that is a pure subset of the `read` resource.

We _can_ ignore the union by recompiling each of the resource names, but that would introduce  Automaton compilation that we don't otherwise need.